internal/controller/postgrescluster/volumes.go

```diff
@@ -122,6 +131,555 @@ func (r *Reconciler) observePersistentVolumeClaims(
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGDataVolume.
+				Directory != "" &&
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGDataVolume.
+				PVCName != "" {
+			pgDataReturn, err = r.reconcileMovePGDataDir(ctx, cluster, moveJobs)
+		}
+
+		if err == nil &&
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGWALVolume != nil &&
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGWALVolume.
+				Directory != "" &&
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGWALVolume.
+				PVCName != "" {
+			pgWALReturn, err = r.reconcileMoveWALDir(ctx, cluster, moveJobs)
+		}
+
+		if err == nil &&
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGBackRestVolume != nil &&
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGBackRestVolume.
+				Directory != "" &&
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGBackRestVolume.
+				PVCName != "" {
+			repoReturn, err = r.reconcileMoveRepoDir(ctx, cluster, moveJobs)
+		}
+		// if any of the 'return early' values are true, return true
+		return pgDataReturn || pgWALReturn || repoReturn, err
+	}
+
+	return false, nil
+}
+
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=create;patch;delete
+
+// reconcileMovePGDataDir creates a Job to move the provided pgData directory
+// in the given volume to the expected location before the PostgresCluster is
+// bootstrapped. It returns any errors and a boolean indicating whether the
+// main control loop should continue or return early to allow time for the job
+// to complete.
+func (r *Reconciler) reconcileMovePGDataDir(ctx context.Context,
+	cluster *v1beta1.PostgresCluster, moveJobs *batchv1.JobList) (bool, error) {
+
+	moveDirJob := &batchv1.Job{}
+	moveDirJob.ObjectMeta = naming.MovePGDataDirJob(cluster)
+
+	// check for an existing Job
+	for i := range moveJobs.Items {
+		if moveJobs.Items[i].Name == moveDirJob.Name {
+			if jobCompleted(&moveJobs.Items[i]) {
+				// if the Job is completed, return as this only needs to run once
+				return false, nil
+			}
+			if !jobFailed(&moveJobs.Items[i]) {
+				// if the Job otherwise exists and has not failed, return and
+				// give the Job time to finish
+				return true, nil
+			}
+		}
+	}
+
+	// at this point, the Job either wasn't found or it has failed, so the it
+	// should be created
+	moveDirJob.ObjectMeta.Annotations = naming.Merge(cluster.Spec.Metadata.
+		GetAnnotationsOrNil())
+	labels := naming.Merge(cluster.Spec.Metadata.GetLabelsOrNil(),
+		naming.DirectoryMoveJobLabels(cluster.Name),
+		map[string]string{
+			naming.LabelMovePGDataDir: "",
+		})
+	moveDirJob.ObjectMeta.Labels = labels
+
+	script := fmt.Sprintf(`echo "Preparing cluster %s volumes for PGO v5.x"
+    echo "pgdata_pvc=%s"
+    echo "Current PG data directory volume contents:" 
+    ls -lh "/pgdata"
+    echo "Now updating PG data directory..."
+    [ -d "/pgdata/%s" ] && mv "/pgdata/%s" "/pgdata/pg%s_bootstrap"
+    rm -f "/pgdata/pg%s/patroni.dynamic.json"
```

Comment:
`patroni.dynamic.json` holds the previous state of the DCS. Since we are migrating the volumes, we want to clear out any obsolete configuration info. I can put a comment saying so.